File: price/eurail.py
from .engine import Engine, Journey
import requests
import datetime as dt

DT_FORMAT = "%Y-%m-%dT%H:%M:%S"

class EurailEngine(Engine):
    def __init__(self):
        self.url = "https://api.timetable.eurail.com/v2/timetable?origin={fromCityId}&destination={toCityId}&timestamp={timestamp}&tripsNumber=5&currency=USD"
        self.station_url = "https://api.timetable.eurail.com/v2/locations?input={query}&results=1"

    def station_to_id(self, station):
        response = requests.get(self.station_url.format(query=station))
        stations = response.json()
        print("#= S_TO_ID", station, " converted to ", stations)
        return stations[0]['id']

    def get_journeys(self, from_city: str, to_city: str, date: dt.datetime) -> list[Journey]:
        # todo currency
        timestamp = date.strftime(DT_FORMAT + ".000Z")

        from_id, to_id = self.station_to_id(from_city), self.station_to_id(to_city)
        url = self.url.format(fromCityId=from_id, toCityId=to_city, timestamp=timestamp)
        trips = requests.get(url).json()

        journeys = [Journey(price=i.get('price', 0), length=self.get_length(i)) for i in trips]
        return journeys

    def get_length(self, trip):
        start = dt.datetime.strptime(trip['departure'], DT_FORMAT)
        end = dt.datetime.strptime(trip['arrival'], DT_FORMAT)

        return (end - start).total_seconds()

File: juliuste-api/src/eurail.js
import strftime from 'strftime';

let DT_FORMAT = "%Y-%m-%dT%H:%M:%S.000Z"
let url = (fromCityId, toCityId, timestamp) => `https://api.timetable.eurail.com/v2/timetable?origin=${fromCityId}&destination=${toCityId}&timestamp=${timestamp}&tripsNumber=5&currency=USD`
let stationUrl = (query) => `https://api.timetable.eurail.com/v2/locations?input=${query}&results=1`

async function stationToId(station) {
    return fetch(stationUrl(station))
        .then(response => response.json())
        .then(result => result[0].id);
}

async function get_journeys(from_city, to_city, date) {
    // todo currency
    // todo date
    let now = new Date();
    let timestamp = strftime(DT_FORMAT, now);

    let from_id = await stationToId(from_city);
    let to_id = await stationToId(to_city);

    return fetch(url(from_id, to_id, timestamp))
        .then(response => response.json())
        .then(result => result.map(trip => {
            let start = new Date(trip.departure);
            let end = new Date(trip.arrival);
            let length = (end - start) / 1000;
            return {
                price: trip.price ?? 0,
                length: length
            }
        }));
}

export default async (req, res) => {
    res.status(200).json({ "journeys": await get_journeys(req.query.origin, req.query.destination, req.query.date) });
}